### ISSUE TITLE
Generate all headers in single template and merge once back

### ISSUE THREAD

**Issue by pofider** [jsreport/jsreport-pdf-utils#10]
The [original request](https://forum.jsreport.net/topic/625/slow-pdf-utils-merge-with-render-for-every-page/3). The idea is to dramatically speed up rendering with dynamic headers.

The main content would run normally.
We would parse the pdf as we do with selected "Render for every page" afterwards.
The merge operation would run single template which could use templating engine "loop" to iterate over parsed pages and render one long pdf.

```html
{{#each $pdf.pages}}
  <div> dynamic header from current page</div>
  <div class='page-break'/>
{{/each}}
```

That would get merged into the original pdf. 
This means there would be only two pdf rendering.

**Comment by pofider** [comment]
I implemented this feature and released it in the "beta" channel .

You can get it using
```
npm install jsreport-pdf-utils@beta
```

Then use flag "merge whole document".

![image](https://user-images.githubusercontent.com/2041960/47581718-8e22ef80-d952-11e8-967e-54a615f0b49e.png)

The header can look like this

```html
{{#each $pdf.pages}}
  {{#if @index}}
    <div style="page-break-before: always;"></div>
  {{/if}}
  <div> 
  dynamic header with page value 
  {{#with (lookup ../$pdf.pages @index)}}
    {{group}}
  {{/with}}
  </div>
{{/each}}
```

and main template

```html
{{#each items}}
{{#if @index}}
<div style='page-break-before: always;'></div>
{{/if}}
<h1>Hello from Page {{this}}</h1>
{{{pdfCreatePagesGroup this}}}
{{/each}}
```

To get even better performance you can install the latest jsreport-chrome-pdf extension which supports chrome instance reusing.
```
npm i jsreport-chrome-pdf@1.2.0
```

and configure reusing chrome instances

```
{
  "extensions": { "chrome-pdf": { "strategy": "chrome-pool", "numberOfWorkers": 3 } } 
}
```

**Comment by bolenti** [comment]
Dear @pofider ,
I have just updated to the beta version, set my main template to _Merge whole documents_ to test this functionality but using _pdfAddPageItem_ and I didn't notice any difference in performance, i.e: with a large a data set the rendering is still slow and timing out.

I have used the code example based on the playground: https://playground.jsreport.net/w/bjrmatos/hiI7sjj~
But I have updated the footer template with the following:

`{{#each $pdf.pages}}`
`  {{#if @index}}`
`    <div style="page-break-before: always;"></div>`
`  {{/if}}`
`  <span class="price" style="position: absolute; bottom: 15px">`
`   {{#with (lookup ../$pdf.pages @index)}}`
`     {{mySum}}`
`   {{/with}}`
`  </span>`
`{{/each}}`

Is there anything I may be missing? If yes, could you please provide a useful example based on the use case presented in the playground URL above?

Many thanks.

**Comment by pofider** [comment]
@bolenti 
Rendering with flag "mergeWholeDocuments" should be significantly faster than with flag "renderForEveryPage". Take a look on this export.
[export (8).zip](https://github.com/jsreport/jsreport-pdf-utils/files/2535481/export.8.zip)
The default variant with "renderForEveryPage" and "Header" merged takes 20s.
If you switch the pdf utils configuratoin to "mergeWholeDocuments" and template "Headers" it should run just 2.5s.

**Comment by bolenti** [comment]
@pofider: Thanks for this example.

Please see my example: [export.zip](https://drive.google.com/file/d/1JXVXY8HGS_PanCdo_-0S6KKbIkARAJqU/view?usp=sharing) attached with two issues:
1) The rendering is slow and would timeout
2) If a reduce the data items to see if things work, the footer is rendered several times on first page only and the same place.

Many thanks.

**Comment by pofider** [comment]
@bolenti 
1. The rendering timeouts even without pdf utils. So it has nothing to do with it. And there is nothing we can do to help because this is the chrome separate task we can't optimize. I don't know if you really need  400 000 rows on the invoice.
2. The footer is rendering values over each other because you defined it that way. Take a look again on my example. I put a page break between each footer/header.

**Comment by bolenti** [comment]
@pofider:
Thanks for your reply
1) Okay I understand the limitation here.
2) Even if I add the page break as in your example (code snippet below), the values in the footer overlap

`  {{#if @index}}
    <div style="page-break-before: always;"></div>
  {{/if}}`

Regards

**Comment by RaphyLi** [comment]
Good feature 👍 

Any ETA for this new feature in docker file?

**Comment by pofider** [comment]
@RaphyLi 
No ETA yet. We will first release it officially in jsreport package during the next week I hope. Then we will see.

**Comment by bjrmatos** [comment]
this feature is published now in `1.2.0`, which is part of jsreport `2.3.0`
